File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Uuid/UUIDTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Foundation;
using Db4objects.Db4o.Query;
using Db4objects.Db4o.Tests.Common.Uuid;
using Sharpen;

namespace Db4objects.Db4o.Tests.Common.Uuid
{
	public class UUIDTestCase : AbstractDb4oTestCase
	{
		private static long storeStartTime;

		private static long storeEndTime;

		public static void Main(string[] args)
		{
			new UUIDTestCase().RunAll();
		}

		public class Item
		{
			public string name;

			public Item(string name_)
			{
				this.name = name_;
			}
		}

		protected override void Configure(IConfiguration config)
		{
			config.ObjectClass(typeof(UUIDTestCase.Item)).GenerateUUIDs(true);
		}

		protected override void Store()
		{
			storeStartTime = Runtime.CurrentTimeMillis();
			Db().Store(new UUIDTestCase.Item("one"));
			Db().Commit();
			storeEndTime = Runtime.CurrentTimeMillis();
			Db().Store(new UUIDTestCase.Item("two"));
		}

		/// <exception cref="System.Exception"></exception>
		public virtual void TestRetrieve()
		{
			Hashtable4 uuidCache = new Hashtable4();
			AssertItemsCanBeRetrievedByUUID(uuidCache);
			Reopen();
			AssertItemsCanBeRetrievedByUUID(uuidCache);
		}

		public virtual void TestTimeStamp()
		{
			IQuery q = NewItemQuery();
			q.Descend("name").Constrain("one");
			UUIDTestCase.Item item = (UUIDTestCase.Item)q.Execute().Next();
			Db4oUUID uuid = Uuid(item);
			long longPart = uuid.GetLongPart();
			long creationTime = TimeStampIdGenerator.IdToMilliseconds(longPart);
			Assert.IsGreaterOrEqual(storeStartTime, creationTime);
			Assert.IsSmallerOrEqual(storeEndTime, creationTime);
		}

		protected virtual void AssertItemsCanBeRetrievedByUUID(Hashtable4 uuidCache)
		{
			IQuery q = NewItemQuery();
			IObjectSet objectSet = q.Execute();
			while (objectSet.HasNext())
			{
				UUIDTestCase.Item item = (UUIDTestCase.Item)objectSet.Next();
				Db4oUUID uuid = Uuid(item);
				Assert.IsNotNull(uuid);
				Assert.AreSame(item, Db().GetByUUID(uuid));
				Db4oUUID cached = (Db4oUUID)uuidCache.Get(item.name);
				if (cached != null)
				{
					Assert.AreEqual(cached, uuid);
				}
				else
				{
					uuidCache.Put(item.name, uuid);
				}
			}
		}

		private Db4oUUID Uuid(object obj)
		{
			return Db().GetObjectInfo(obj).GetUUID();
		}

		private IQuery NewItemQuery()
		{
			return NewQuery(typeof(UUIDTestCase.Item));
		}
	}
}

File: db4oj.tests/src/com/db4o/db4ounit/common/uuid/UUIDTestCase.java
/* Copyright (C) 2004 - 2008  Versant Inc.  http://www.db4o.com */

package com.db4o.db4ounit.common.uuid;

import com.db4o.*;
import com.db4o.config.*;
import com.db4o.ext.*;
import com.db4o.foundation.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.*;

public class UUIDTestCase extends AbstractDb4oTestCase {
    
    private static long storeStartTime;
    
    private static long storeEndTime;
	
	public static void main(String[] args) {
		new UUIDTestCase().runAll();
	}
	
    public static class Item {
        
        public String name;

        public Item(String name_) {
            this.name = name_;
        }
        
    }
	
	protected void configure(Configuration config) {
		config.objectClass(Item.class).generateUUIDs(true);
	}

	protected void store() {
	    storeStartTime = System.currentTimeMillis();
		db().store(new Item("one"));
		db().commit();
		storeEndTime = System.currentTimeMillis();
		db().store(new Item("two"));
	}

	public void testRetrieve() throws Exception {
		Hashtable4 uuidCache = new Hashtable4();
		assertItemsCanBeRetrievedByUUID(uuidCache);
		reopen();
		assertItemsCanBeRetrievedByUUID(uuidCache);
	}
	
	public void testTimeStamp(){
	    Query q = newItemQuery();
	    q.descend("name").constrain("one");
	    Item item = (Item) q.execute().next();
	    Db4oUUID uuid = uuid(item);
	    long longPart = uuid.getLongPart();
	    long creationTime = TimeStampIdGenerator.idToMilliseconds(longPart);
	    Assert.isGreaterOrEqual(storeStartTime, creationTime);
	    Assert.isSmallerOrEqual(storeEndTime, creationTime);
	}

    protected void assertItemsCanBeRetrievedByUUID(Hashtable4 uuidCache) {
        Query q = newItemQuery();
        ObjectSet objectSet = q.execute();
        while (objectSet.hasNext()) {
            Item item = (Item) objectSet.next();
            Db4oUUID uuid = uuid(item);            
            Assert.isNotNull(uuid);
            Assert.areSame(item, db().getByUUID(uuid));
            final Db4oUUID cached = (Db4oUUID) uuidCache.get(item.name);
            if (cached != null) {
                Assert.areEqual(cached, uuid);
            } else {
                uuidCache.put(item.name, uuid);
            }
        }
    }

    private Db4oUUID uuid(Object obj) {
        return db().getObjectInfo(obj).getUUID();
    }

    private Query newItemQuery() {
        return newQuery(Item.class);
    }

}
